lib/git/base.rb

```diff
@@ -92,18 +98,37 @@ def initialize(options = {})
       @index = options[:index] ? Git::Index.new(options[:index], false) : nil
     end
     
-    # changes current working directory for a block
-    # to the git working directory
+    # changes current working directory for a block to the git working directory.
+    #
+    # Note: If we can fork() or spawn(), Dir.chdir will happen in a new process
+    # otherwise, we will use a mutex to prevent threading errors
+    # See https://github.com/ruby-git/ruby-git/issues/355 for more info
     #
     # example
     #  @git.chdir do 
     #    # write files
     #    @git.add
     #    @git.commit('message')
     #  end
-    def chdir # :yields: the Git::Path
-      Dir.chdir(dir.path) do
-        yield dir.path
+    def chdir(&block) # :yields: the Git::Path
+      chdir_block = Proc.new do
+        Dir.chdir(dir.path) do
+          block.call(dir.path)
+        end
+      end
+
+      if Process.respond_to?(:fork)
+        # Forking this process so that we can be threadsafe
+        pid = Process.fork do
+          chdir_block.call
+        end
+        Process.wait(pid)
+      else
+        # Windows and NetBSD 4 don't support fork()
```

Comment:
Interesting anecdote, I wasn't aware of NetBSD 4 here. But OTOH, it's past EOL already. Maybe we can drop that reference and instead mention JRuby (which is actively used by many), since it also has problems with `fork()` on the JVM.